libcalico-go/lib/packedmap/packedmap.go

```diff
@@ -0,0 +1,160 @@
+// See also MakeCompressedJSON and MakeDedupedCompressedJSON.
+type RawPackedMap[K comparable, VIn, VPacked any] struct {
+	encoder Encoder[VIn, VPacked]
+	m       map[K]VPacked
+}
+
+func Make[K comparable, VIn, VPacked any](encoder Encoder[VIn, VPacked]) RawPackedMap[K, VIn, VPacked] {
+	return RawPackedMap[K, VIn, VPacked]{
+		encoder: encoder,
+		m:       make(map[K]VPacked),
+	}
+}
+
+// MakeCompressedJSON returns a new map that stores JSON-encoded values,
+// compressed with snappy compression.
+func MakeCompressedJSON[K comparable, V any]() Map[K, V] {
+	return Map[K, V]{
+		Make[K, V, string](
+			SnappyEncoderWrapper[V]{
+				encoder: JSONEncoder[V]{},
+			},
+		),
+	}
+}
+
+// Map is a convenience wrapper around RawPackedMap that assumes string as the
+// packed type.
+type Map[K comparable, V any] struct {
+	RawPackedMap[K, V, string]
+}
+
+// MakeDedupedCompressedJSON returns a new map that stores JSON-encoded values,
+// compressed with snappy compression and then deduped using the unique package.
+func MakeDedupedCompressedJSON[K comparable, V any]() Deduped[K, V] {
+	pm := Make[K, V, unique.Handle[string]](
+		DedupingEncoderWrapper[V, string]{
+			encoder: SnappyEncoderWrapper[V]{
+				encoder: JSONEncoder[V]{},
+			},
+		},
+	)
+	return Deduped[K, V]{pm}
+}
+
+// Deduped is a convenience wrapper around RawPackedMap that assumes
+// unique.Handle[string], as used by the DedupingEncoderWrapper.
+type Deduped[K comparable, V any] struct {
+	RawPackedMap[K, V, unique.Handle[string]]
+}
+
+func (pm RawPackedMap[K, VIn, VPacked]) Set(key K, val VIn) {
+	pm.m[key] = pm.encoder.Pack(val)
+}
+
+func (pm RawPackedMap[K, VIn, VPacked]) Get(key K) (val VIn, ok bool) {
+	packed, ok := pm.m[key]
+	if !ok {
+		return
+	}
+	val = pm.encoder.Unpack(packed)
+	return
+}
+
+func (pm RawPackedMap[K, VIn, VPacked]) Delete(key K) {
+	delete(pm.m, key)
+}
+
+func (pm RawPackedMap[K, VIn, VPacked]) Len() int {
+	return len(pm.m)
+}
+
+// Encoder is a type that can pack and unpack values for use in a packed map.
+type Encoder[VIn, VOut any] interface {
+	Pack(val VIn) VOut
+	Unpack(val VOut) VIn
+}
+
+// DedupingEncoderWrapper is an Encoder that wraps another Encoder and dedupes
+// the packed values using the unique package.
+type DedupingEncoderWrapper[VIn any, VOut comparable] struct {
+	encoder Encoder[VIn, VOut]
+}
+
+func (p DedupingEncoderWrapper[VIn, VOut]) Pack(val VIn) unique.Handle[VOut] {
+	packed := p.encoder.Pack(val)
+	return unique.Make(packed)
+}
+
+func (p DedupingEncoderWrapper[VIn, VOut]) Unpack(packed unique.Handle[VOut]) VIn {
+	return p.encoder.Unpack(packed.Value())
+}
+
+// SnappyEncoderWrapper is an Encoder that wraps another Encoder and compresses
+// the packed values using snappy.
+type SnappyEncoderWrapper[V any] struct {
+	encoder JSONEncoder[V]
+}
+
+func (p SnappyEncoderWrapper[V]) Pack(val V) string {
+	buf := p.encoder.Pack(val)
+	var arr [1024 * 16]byte
+	packed := snappy.Encode(arr[:], []byte(buf))
```

Comment:
packed would be a subslice of arr if arr was large enough. Does it mean that it always takes 16k at least?

Reply:
Tweaked it to re-use the same cached buffer over and over.  That will force the compiler to allocate the string, as desired.  (Right now, the go compiler does allocate but I couldn't find anything in the language spec to say it always will.)